pkg/virt-handler/selinux/context_executor.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package selinux
+
+import (
+	"fmt"
+	"os/exec"
+	"runtime"
+	"syscall"
+
+	"github.com/opencontainers/selinux/go-selinux"
+)
+
+type SELinuxContextExecutor struct {
+	cmdToExecute  *exec.Cmd
+	desiredLabel  string
+	originalLabel string
+	pid           int
+}
+
+func NewSELinuxContextExecutor(pid int, cmd *exec.Cmd) (*SELinuxContextExecutor, error) {
+	desiredLabel, err := getSelinuxLabelForPID(pid)
+	if err != nil {
+		return nil, err
+	}
+	originalLabel, err := getOriginalSelinuxLabel()
+	if err != nil {
+		return nil, err
+	}
+	return &SELinuxContextExecutor{
+		pid:           pid,
+		cmdToExecute:  cmd,
+		desiredLabel:  desiredLabel,
+		originalLabel: originalLabel,
+	}, nil
+}
+
+func (ce SELinuxContextExecutor) Execute() error {
+	if isSELinuxEnabled() {
+		if err := ce.setDesiredContext(); err != nil {
+			return err
+		}
+	}
+	defer ce.resetContext()
+
+	preventFDLeakOntoChild()
+	if err := ce.cmdToExecute.Run(); err != nil {
+		return fmt.Errorf("failed to execute command in launcher namespace %d: %v", ce.pid, err)
+	}
+	return nil
+}
+
+func (ce SELinuxContextExecutor) setDesiredContext() error {
+	runtime.LockOSThread()
+	if err := selinux.SetExecLabel(ce.desiredLabel); err != nil {
+		return fmt.Errorf("failed to switch selinux context to %s. Reason: %v", ce.desiredLabel, err)
+	}
+	return nil
+}
+
+func (ce SELinuxContextExecutor) resetContext() error {
+	err := selinux.SetExecLabel(ce.originalLabel)
+	runtime.UnlockOSThread()
```

Comment:
Will it be safer to use `defer` here? You seem to do this even if the label operations fails, so maybe it should also be done in case it panics.
```
defer runtime.UnlockOSThread()
return selinux.SetExecLabel(ce.originalLabel)
```